GPUVerifyVCGen/GPUVerifier.cs

```diff
@@ -1740,10 +1740,15 @@ private void GenerateBarrierImplementation(Procedure barrierProcedure)
 
             Debug.Assert(p1 != null);
             Debug.Assert(p2 != null);
-            Debug.Assert(localFence1 != null);
-            Debug.Assert(localFence2 != null);
-            Debug.Assert(globalFence1 != null);
-            Debug.Assert(globalFence2 != null);
+
+            // the assertions are applicable only to block-level barriers
+            if (IsBlockBarrier(barrierProcedure))
+            {
+                Debug.Assert(localFence1 != null);
+                Debug.Assert(localFence2 != null);
+                Debug.Assert(globalFence1 != null);
+                Debug.Assert(globalFence2 != null);
+            }
```

Comment:
I will review the conditions for  **divergenceCondition** and **ifGuard** and come back on them.

Reply:
I added the grid barrier check at both locations. What I noticed is that the statements in both the if-blocks are executed only if uniformity analysis by Boogie is disabled. I was able to disable it by using **/noUniformityAnalysis** option when calling **GPUVerifyVCGen**. However, I don't see an option of disabling this through **GPUVerify.py**
I am not familiar with Boogie's uniformity analysis, can you please review these changes to make sure I haven't inadvertently broken something else?

Reply:
When did you see the problem? With specific tests, or everywhere? If with specific tests, which ones? Note that a further data point might be obtained here by adding the test I suggested you should add.

Briefly: uniformity analysis check whether on a path from kernel entry to a barrier there's a test that can make a subset of the threads not reach the barrier (so this will typically involve the thread id). It might be that we need to do some extra work here, as for grid barriers the block id is also relevant. I will need to dig into the code a bit to see what is going on. It's been a while since I last looked at this.

Reply:
I added this test case and I see the issue now. The test case is not failing as excepted. That's because the divergence check is happening only for the threads within the same group as you mentioned. I will review that further and make the necessary code changes.

One thing that I noticed from the CUDA Cooperative Groups documentation is that they specify that the behavior is undefined when only a subset of threads reach the thread-level barrier (https://docs.nvidia.com/cuda/cuda-c-programming-guide/index.html#thread-block-group-cg) but they don't explicitly mention this for grid-level barriers. But I think it's fair to assume that the behavior is undefined when only a subset of threads reach the grid-level barrier.

Reply:
Thanks for adding the test case. I'll also try to have a look at this over the weekend.

Reply:
To fix the divergence issue, what you want to do is rewrite `IdentifySafeBarriers`:
```
        private void IdentifySafeBarriers()
        {
            var uniBlock = DoUniformityAnalysis(new List<Variable> { IdX, IdY, IdZ });
            var uniGrid = DoUniformityAnalysis(new List<Variable> { IdX, IdY, IdZ, groupIdX, groupIdY, groupIdZ });
            foreach (var b in barrierProcedures)
            {
                if ((IsBlockBarrier(b) && uniBlock.IsUniform(b.Name)) || (IsGridBarrier(b) && uniGrid.IsUniform(b.Name)))
                {
                    b.AddAttribute("safe_barrier", new object[] { });
                }
            }
        }
```

Reply:
thank you, I will take a look into this.